emscriptenStdenv: document .emscriptencache (#181363)
NixOS/nixpkgs: doc/languages-frameworks/emscripten.section.md

## Declarative usage {#declarative-usage}

Building emscripten packages requires some special considerations which developers should be aware of. See [`pkgs/top-level/emscripten-packages.nix`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/emscripten-packages.nix) for details and examples.

* `pkgs.buildEmscriptenPackage` sets `docheck = true` by default, which requires that each emscriptenPackage have a `checkPhase`.
* Use `export EMCC_DEBUG=2` from within a emscriptenPackage's `phase` to get more detailed debug output what is going wrong.

> * The ``~/.emscripten`` cache requires us to set `HOME=$TMPDIR` in individual phases. This makes compilation slower but also makes it more deterministic. by default the emscripten cache location is set to a new writable `.emscriptencache` directory via the environment variable `EM_CACHE`, because the systemwide default directory is unwritable in the nix store. If overriding the `configurePhase` while creating an emscripten package, this step will need to be included manually  as well.

Reviewer: Why is that not done in buildEmscriptenPackage ?

Author: I should start by acknowledging that I'm still new enough to nixpkgs internals that the distinction between the emscriptenStdenv and buildEmscriptenPackages, as defined by https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/em-modules/generic/default.nix, is a little difficult for me to totally understand. That said, do you mean that there's a way to force those cache settings to happen even if a new `configurePhase` is defined? I assumed that once a user/derivation starts overriding things then it's their responsibility to recreate all the default workarounds, so that nothing else magical is happening behind the scenes once the user has decided to pull back those abstractions.

Reviewer 2: Maybe we can set `EM_CACHE = "$TMP";` in `mkDerivation`?

Reviewer: > That said, do you mean that there's a way to force those cache settings to happen even if a new configurePhase is defined?

There are multiple ways. One would be to defined it in an earlier hook for emscripten that is automatically run.

* Use `export EMCC_DEBUG=2` from within a emscriptenPackage's `phase` to get more detailed debug output what is going wrong.

> * ~/.emscripten cache is requiring us to set `HOME=$TMPDIR` in individual phases. This makes compilation slower but also makes it more deterministic.

Reviewer 3: ```suggestion
* `~/.emscriptencache` is requires setting `HOME=$TMPDIR` in individual phases. This makes compilation slower but also makes it more deterministic.
```

> * ~/.emscripten cache is requiring us to set `HOME=$TMPDIR` in individual phases. This makes compilation slower but also makes it more deterministic.

Reviewer 3: ```suggestion
* `HOME=$TMPDIR` is set in phases that access `~/.emscripten`. This precludes reusing build artifacts across builds, but also makes compilation more deterministic.
```

> * Additionally, during the `configurePhase` of `emscriptenStdenv`, a local (writable) directory is created at `$(pwd)/.emscriptencache`, and the environment variable `EM_CACHE` is set to that path. If overriding the `configurePhase` for a derivation, this step will need to be included manually as well.

Reviewer 3: What does `$(pwd)` refer to here?

Author: It's certainly possible I should be phrasing it better -- I took it perhaps too literally from https://github.com/NixOS/nixpkgs/pull/172207/files#diff-de4924a70c218dfafd242eda61aaffb33790c7b14b90eed4bf9e3c3aaf2db37aR28-R30, but in reality it's just in whatever directory `emconfigure` would be running inside. Perhaps it's okay to just remove the `$(pwd)/`, since that's more bash-related.

Reviewer 3: Ah, seeing the code helps a lot. That note mostly describes what the code does at a level of detail which equals the code, which I find problematic. Documentation should describe what the user should do to succeed, and only describe the code as far as necessary to understand why. Also, multiple packages set `EM_CACHE`, so that part is not correct.

I suggest removing the new line and changing the above one (see https://github.com/NixOS/nixpkgs/pull/181363/files/08d63f8b6907b679d473f068efd49e567f8b0858#r921869466).

```suggestion
```

I cannot make suggestions for the other lines. Could you please
- delete l. 39-44
- add an introductory sentence to the bullet list, something to the effect of "building emscripten packages requires some special considerations to be aware of, see `pkgs/top-level/emscripten-packages.nix` for details" (with a permalink to the current master commit)
- make the `doCheck = true` line part of the list instead of a sentence and reword to "`pkgs.buildEmscriptenPackage` sets `doCheck = true` by default, which requires each `emscriptenPackage` to have a `checkPhase`.

Author: Updated with latest force push @fricklerhandwerk. Let me know if it makes more sense now!
